pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -190,7 +192,11 @@ public void addSegment(ImmutableSegmentImpl segment, @Nullable ThreadSafeMutable
       if (queryableDocIds == null && _deleteRecordColumn != null) {
         queryableDocIds = new ThreadSafeMutableRoaringBitmap();
       }
-      addOrReplaceSegment(segment, validDocIds, queryableDocIds, recordInfoIterator, null, null);
+      if (_tableUpsertMetadataManager.isPreloadingSegment(segmentName)) {
```

Comment:
I feel it is cleaner if we add `preloadSegment(ImmutableSegment segment)` into `PartitionUpsertMetadataManager` and directly call it in `RealtimeTableDataManager.handleUpsert()`. This way we don't need to pass `TableUpsertMetadataManager` into `PartitionUpsertMetadataManager`, and during preload, we can perform some extra checks such as snapshot exist etc.

Reply:
This waiting logic can be removed as discussed in another comment. So likely no changes on this class.

as in this PR, the TableUpsertMetadataMgr does preload and also checks snapshot existence. I’ll think about if adding that preloadSegment() would simplify the code further.

Reply:
Currently the snapshot existence check is done in table upsert mgr’s preloadSegments method. But I’ll think more about the idea proposed here.

Reply:
hmm.. after some thoughts, I could not figure out how to make the changes you proposed.

Basically, if we call `preloadSegment(ImmutableSegment segment)` as part of handleUpsert(), which is part of addSegment(ImmutableSegment segment))` of `RealtimeTableDataManager`. I think that's too late for the snapshot existence check, because the segment is already fully loaded as the `ImmutableSegment` object is already created. The logic in current PR can skip the segment if it doesn't have the snapshot file.

Reply:
updated. thanks for idea 👍